compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+    Type** castTo) {
+
+  if(sym && !(toFnSymbol(sym) || toArgSymbol(sym) || toTypeSymbol(sym))) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar = NULL;
+
+    Vec<SymExpr*>* defs = defMap.get(sym);
+    Vec<SymExpr*>* uses = useMap.get(sym);
+
+    if(defs && defs->n == 1 && uses && uses->n == 1) { // check def-use counts
+      SymExpr* se = defs->first();
+      defPar = se->parentExpr;
+
+      //defPar has to be a move without any coercion
+      CallExpr* ce = toCallExpr(defPar);
+      if(ce) {
+        if(ce->isPrimitive(PRIM_MOVE)) {
+          Type* lhsType = ce->get(1)->typeInfo();
+          Type* rhsType = ce->get(2)->typeInfo();
+          if(lhsType == rhsType) {
+            if(! canPrimMoveCreateCommunication(ce)) {
+              if(! (lhsType->symbol->hasFlag(FLAG_EXTERN))){
+                if(!lhsType->symbol->hasFlag(FLAG_ATOMIC_TYPE)){
+                  //at this point we now that def is fine
+                  def = ce->get(2);
+
+                  //now check if we need to case it when we move it
+                  if(CallExpr* defCe = toCallExpr(def)) {
+                    if(defCe->isPrimitive() &&
+                        isIntegerPromotionPrimitive(defCe->primitive->tag)) {
+                      if(requiresCast(lhsType)) {
+                        *castTo = lhsType;
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+
+      if(def) {
+        *defOut = def;
+        // we have def now find where the value is used
+        SymExpr* se = uses->first();
+        usePar = se->parentExpr;
+        if(CallExpr* ce = toCallExpr(usePar)) {
+          if( !(ce->isPrimitive(PRIM_ADDR_OF) ||
+                ce->isPrimitive(PRIM_ARRAY_GET) ||
+                ce->isPrimitive(PRIM_GET_MEMBER) ||
+                ce->isPrimitive(PRIM_DEREF) ||
+                ce->isPrimitive(PRIM_GET_MEMBER_VALUE) ||
+                (ce->isPrimitive(PRIM_MOVE) && 
+                 ce->get(1)->typeInfo() !=
+                 ce->get(2)->typeInfo()))) {
+            use = se;
+          }
+        }
+      }
+      else {
+        return false;
+      }
+      if(use) {
+        *useOut = use;
+      }
+      else {
+        return false;
+      }
+
+      //here I have my valid use and def
+      //so far we checked specific cases for them individually, now check if
+      //there is anything wrong with them as a pair
+
+      //this issue feels too specific, maybe there is
+      //safer/better/more general way of doing this check
+      //
+      //for reference test that caused this was:
+      //test/modules/standard/FileSystem/bharshbarg/filer
+      //
+      //The issue seemed to be yielding string from an iterator
+      if(CallExpr* useParentCe = toCallExpr(usePar)) {
+        if(useParentCe->isPrimitive(PRIM_FTABLE_CALL)) {
```

Comment:
Is this issue actually from strings (and all records, generally) that have blank argument intent being passed as pointers, but not marked as ref intent at this point?